### PR TITLE
Added a function/context manager and environment variables to enable/disable dispatching and set backend/backend priority

### PR DESCRIPTION
**Pull request by Schefflera-Arboricola**
## Description

previous discussions at : https://github.com/betatim/scikit-image/pull/1

## Existing behaviour - [PR#7520](https://github.com/scikit-image/scikit-image/pull/7520)

- The function call is dispatched to the first backend, in the alphabetically ordered list of backends, which implements the algorithm. 
- Dispatching is enabled by default. 
- `SKIMAGE_NO_DISPATCHING` environment variable is set to `1` to disable it. 

## New behaviour added in this PR

- The default backend priority is still same, but now a user can not set their own priority.
- Dispatching is disabled by default.
- We have two environment variables now:
    - `SKIMAGE_DISPATCHING` : stores if the dispatching is enabled or disabled
    - `SKIMAGE_BACKEND_PRIORITY`: stores the backend or backend priority
- Added `set_backends(*backends, dispatch=True/False)` context manager to enable/disable dispatching and set backend/backend priority. And if both environment variables and `set_backends` are used then preference will be given to values in `set_backends` over the environ var. (refer: https://github.com/betatim/scikit-image/pull/1#discussion_r1951236436)

For usage, see the updated user-facing docs (`backends.rst` in this PR)

# Other changes

- Added user-facing backend docs
- renamed :
    - `all_backends()` --> `all_backends_with_eps_combined()`
    - key `"implementation"` --> `"skimage_backends_ep_obj"`
- improved docstring of all the backend functions to follow the best practices and syntax
- covered corner cases
- removed unnecessary `try-except` in `all_backends_with_eps_combined` as the second entry-point was made mandatory.
- improved dispatching developer facing docs

ToDo:
- [ ] Add tests for `set_backends`

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpydoc/blob/main/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

For maintainers and optionally contributors, please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) on how to document this PR for the release notes.

```release-note
...
```

### CODE REVIEW

**Comment by stefanv** [comment]
It would be good to rebase this PR to get rid of the long history, if that's not too much of a hassle.

File: doc/source/development/dispatching.rst [hunk 1]
```diff
@@ -72,18 +75,21 @@ only want to handle calls where the input arrays are of a particular type or siz
...
 ``get_implementation(name)`` function is called to get the implementation. This should
 return the backend function that implements the behaviour of the function ``name`` in scikit-image.
-The ``name`` parameter will contain the public module name and the function name separated by a
-colon. For example, the ``name`` for the ``canny`` function from the ``feature`` module would
-be ``skimage.feature:canny``.
 
 Once the implementation has been retrieved from the backend, it will be called with the
 arguments the user provided and it is expected to return the result of the computation.
 
+Note that we don't check the passed in args and kwargs when calling the backend
```
[the remark this answers is not in the document]
Comment (Schefflera-Arboricola) [reply doc/source/development/dispatching.rst]: added a note to wrap the function calls with additional backend-specific args in the `set_backends` context manager. here--> https://github.com/scikit-image/scikit-image/pull/7727/commits/16f045b5d33621cb4af723c2ea022755eb0bb99e . LMK if this addresses you comment.

But, even in the `set_backends` context, the user cannot choose whether to fall-back to scikit-image or not, so, pls LMK if it would be a good idea to add `fallback` in `set_backends`, like :
```python
set_backends("backend1", "backend2",... , dispatch=True, fallback=False)
```

File: doc/source/user_guide/backends.rst [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+Scikit-image offers an API that enables dispatching algorithms to specific backends,
+allowing users to run alternative implementations outside of scikit-image while maintaining
+the same user interface. These backend packages may provide optimized versions of scikit-image
+algorithms, support different array types beyond NumPy, be implemented in other languages
+like C++ or Rust, or be tailored for specific hardware such as GPUs.
+
+Using backends
+--------------
+
+Firstly, you need to install the scikit-image backend package you need to use.
```
[the remark this answers is not in the document]
Comment (Schefflera-Arboricola) [reply doc/source/user_guide/backends.rst]: I avoided mentioning a specific backend to remain neutral. LMK if adding something like this to the docs would be a good idea: 

```
First, install the scikit-image backend you want to use::

    $ pip install git+https://github.com/Schefflera-Arboricola/skimage-j4f.git@main

```

File: doc/source/user_guide/backends.rst [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+By default, the backend dispatching is **disabled**. To enable and customize backend dispatching, you
```
[the remark this answers is not in the document]
Comment (Schefflera-Arboricola) [reply doc/source/user_guide/backends.rst]: As discussed in the last meeting, we agreed upon to make users enable the dispatching explicitly so that they are aware of it and they don't run into any unexpected behaviours. Also I have added the `set_backends` function/context manager to enable/disable dispatching and set backend/backend priority. More info in the updated PR description.

File: skimage/__init__.pyi [hunk 4]
```diff
@@ -53,3 +53,5 @@ from .util.dtype import (
 )
 from .util.lookfor import lookfor
 from .data import data_dir
+
+from .util._backends import set_backends
```
Comment (Schefflera-Arboricola) [inline skimage/__init__.pyi:57]: I'm not sure if this is the right place to put it.

File: doc/source/user_guide/backends.rst [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+like C++ or Rust, or be tailored for specific hardware such as GPUs.
+
+Using backends
+--------------
+
+First, install the scikit-image backend you want to use. By default, the backend
+dispatching is **disabled**. To enable it you can either:
+
+- set the environment variable ``SKIMAGE_DISPATCHING = "True"``, or
+- use the ``skimage.set_backends(dispatch=True)`` function in your script.
```
Comment (stefanv) [inline doc/source/user_guide/backends.rst:24]: Unlike with environment variables, we have the ability to express ourselves via a more nuanced API in code.

There are two thoughts we'd like to express: enable / disable dispatching and, if you dispatch, what should the order be?

In terms of environment variables, it's clearest to have:

```
SKIMAGE_DISPATCH = 1  # auto backend prioritization
```

or

```
SKIMAGE_DISPATCH = 1
SKIMAGE_BACKENDS = a,b,c  # explicit prioritization
```

But, in code we could easily write:

```
skimage.use_dispatch()  # enable dispatching, auto backend prioritization
skimage.use_dispatch(backends=('a', 'b', 'c'))  # explicit prioritization
```

Comment (stefanv) [reply doc/source/user_guide/backends.rst:24]: I think something we should discuss as a team is whether we want to encourage the use of `skimage.use_dispatch` (or `skimage.set_backends`, whatever the context manager is called) at all. Does the user truly benefit from this level of fine-grained control? Are they likely to use it? Won't they almost always want to switch whole-sale to the faster backend? That is my guess, and if so there is little advantage in people sprinkling these statements throughout code.

Comment (Schefflera-Arboricola) [reply doc/source/user_guide/backends.rst:24]: right now, this is how `set_backends` behave:

```
skimage.set_backends(dispatch=True)  # enable dispatching, auto backend prioritization
skimage.set_backends('a', 'b', 'c')  # will only set the backend priority and the `dispatch` will remain set to False
skimage.set_backends('a', 'b', 'c', dispatch=True) # explicit prioritization with dispatching enabled
```

The backend priority and the value of `dispatch` in the lastest instance of `set_backends` is considered.

Comment (Schefflera-Arboricola) [reply doc/source/user_guide/backends.rst:24]: replying to https://github.com/scikit-image/scikit-image/pull/7727#discussion_r2002635105 :

> Won't they almost always want to switch whole-sale to the faster backend? That is my guess, and if so there is little advantage in people sprinkling these statements throughout code.

I wasn't imagining a user using the `set_backends` a lot in their script. I only see them using it once in the beginning of their code. I think the motivation, that we discussed, of adding a function/context was that the setting of dispatching will be in the script, so then the same script will not behave differently in different environments based on environment variables' values. And environment variables were beneficial to set dispatching for a bunch of scripts together-- because you said in the last meeting that sometimes people have "pipelines" and various code files and environment variables could be useful in that case. Please LMk if my understanding is correct.

Or maybe, by "fine-grained control" you are trying to say that a user might only want to use one backend and they can just set the `SKIMAGE_NO_DISPATCHING` env var and run their script(s) and backend_priority might not be used a lot. LMK if that is what you are thinking. Thanks!

Comment (betatim) [reply doc/source/user_guide/backends.rst:24]: A few thoughts:

I think allowing people to use environment variables and code to enable/disable/configure dispatching makes sense. I'm not yet convinced you need a context manager to turn it on/off mid script. It is likely people will need/want it but I'd wait with adding it until a user comes with a use-case before actually adding it.

We can re-use `skimage.use_dispatch` as a context manager later. Until that is supported I would add a note saying that calling `use_dispatch()` more than once in your program is "off label use". It might work, it might not work, it might delete your hard drive - following Python's "consenting adults" idea I would not put any technical effort into preventing people from doing it.

I would stick to using one environment variable name, probably `SKIMAGE_DISPATCH` instead of having two (we also need to change the `SKIMAGE_NO_DISPATCHING` that is currently in the code):
```
SKIMAGE_DISPATCH = 1
SKIMAGE_DISPATCH = 0  # new way of saying SKIMAGE_NO_DISPATCHING=1
SKIMAGE_DISPATCH = a,b,c  # explicit prioritization
SKIMAGE_DISPATCH=  # setting env var to no value - I would not allow this
```

I prefer an API that allows users to write the following:
```python
# Enable dispatching (overrides env variable value!?)
skimage.use_dispatch()
# Disable dispatching
skimage.use_dispatch(False)
# Enable dispatching and specify priority of backends
skimage.use_dispatch(backends=("foo", "a", "b", "c"))
skimage.use_dispatch(backends=["foo", "a", "b", "c"])
# A bit awkward maybe but I think important/for completeness
# these are ways to disable dispatching
skimage.use_dispatch(backends=False)
skimage.use_dispatch(backends=[]). # allow people to pass an empty sequence, which is False-y
skimage.use_dispatch(backends=tuple())
# And this is a legal way to enable dispatching
skimage.use_dispatch(backends=True)
```

Except for `skimage.use_dispatch()` I think you could write all of these as `skimage.set_config(backends=...)` to reduce the (minimal) duplication of having backend/dispatch in the function and the argument name. But that also opens the door to other configuration settings `skimage.set_config(foobar=...)` which is a much bigger discussion and maybe too big?

We need to define what the order of priority is when people use env vars and code. I could argue for either order (env wins over code or code wins over env) - what do others think? What is less surprising? What is the "normal" order of precedence in other libraries/tools?

**Comment by matthew-brett** [comment]
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
